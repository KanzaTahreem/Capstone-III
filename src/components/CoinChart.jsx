import React from 'react';
import { useSelector } from 'react-redux';
import { Chart } from 'react-charts';

const CoinChart = () => {
  const chartData = useSelector((state) => state.chart.dataArray);

  const data = React.useMemo(
    () => {
      const mappedPrices = chartData.prices?.map((price, index) => [
        index === chartData.prices.length - 1 ? 'Today' : `${chartData.prices.length - index}d`,
        price[1],
      ]);
      return [
        {
          label: 'Series 1',
          data: mappedPrices || [],
        },
      ];
    },
    [chartData],
  );

  const axes = React.useMemo(
    () => [
      { primary: true, type: 'ordinal', position: 'bottom' },
      { type: 'linear', position: 'left' },
    ],
    [],
  );

  return (
    <div
      style={{
        width: '80vw',
        height: '400px',
      }}
    >
      { chartData.prices && chartData.prices.length ? <Chart data={data} axes={axes} /> : 'Loading...'}
    </div>
  );
};

export default CoinChart;
