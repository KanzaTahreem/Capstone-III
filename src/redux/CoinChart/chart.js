export const FETCH_CHART = 'cryptoData/crypt/FETCH_CHART';
export const SET_LOADING = 'cryptoData/team/SET_LOADING';

const initialState = {
  dataArray: [],
  isLoading: false,
};

export const fetchChart = (payload) => ({
  type: FETCH_CHART,
  payload,
});

export const setLoading = (payload) => ({
  type: SET_LOADING,
  payload,
});

const chartReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_CHART:
      return {
        ...state, dataArray: action.payload,
      };

    case SET_LOADING:
      return {
        ...state, isLoading: action.payload,
      };

    default:
      return state;
  }
};

export default chartReducer;
