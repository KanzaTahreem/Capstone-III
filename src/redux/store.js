import { configureStore } from '@reduxjs/toolkit';
import coinsReducer from './AllCoins/coins';
import chartReducer from './CoinChart/chart';

const store = configureStore({
  reducer: {
    crypto: coinsReducer,
    chart: chartReducer,
  },
});

export default store;
