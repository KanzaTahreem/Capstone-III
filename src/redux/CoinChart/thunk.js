import { createAsyncThunk } from '@reduxjs/toolkit';
import { fetchChart, FETCH_CHART } from './chart';

const fetchChartFromApi = createAsyncThunk(FETCH_CHART, async (coinId, thunkApi) => {
  const response = await fetch(`https://cryptfacts.up.railway.app/v4/details/chart/${coinId}`, {
    method: 'GET',
  });

  const responseJSON = await response.json();
  thunkApi.dispatch(fetchChart(responseJSON));
  return responseJSON;
});

export default fetchChartFromApi;
