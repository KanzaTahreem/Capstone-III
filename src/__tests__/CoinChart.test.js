import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../redux/store';
import CoinChart from '../components/CoinChart';

it('renders correctly', () => {
  const rocket = renderer
    .create(<Provider store={store}><CoinChart /></Provider>)
    .toJSON();
  expect(rocket).toMatchSnapshot();
});
