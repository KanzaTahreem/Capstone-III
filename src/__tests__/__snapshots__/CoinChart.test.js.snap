// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "height": "400px",
      "width": "80vw",
    }
  }
>
  Loading...
</div>
`;
